### PR TITLE
Inference for timl

### PR DESCRIPTION
**Pull request by ivanzvonkov**
Part of [#4 ](https://github.com/nasaharvest/timl/issues/4)

This PR introduces a stand alone inference class which can be used to make predictions with a ckpt or jit model.
This inference class will make it possible to make predictions inside a torchserve server without a lot of dependencies (you'll only need the inference class).

I've left `# TODO` comments where some discussion would be useful.

### CODE REVIEW

File: crop_classification/inference.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    def __init__(self, model, device: Optional[torch.device] = None):
+        self.model = model
+        self.device = device
+
+        # TODO: normalizing_dict is property of the model in crop-mask should this be the same here?
+        self.normalizing_dict: Dict[str, np.ndarray] = None
+
+        # TODO: batch size is a property of the model in crop-mask should this be the same here?
+        self.batch_size: int = 64
```
Comment (gabrieltseng) [inline crop_classification/inference.py]: I think it should be an argument to `run`, since its a property of how the predictions are made (rather than of the model itself)

File: crop_classification/inference.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    """
+    Class for running inference using either a pytorch checkpoint model or
+    a pytorch jit model on a single tif file.
+    """
+
+    def __init__(self, model, device: Optional[torch.device] = None):
+        self.model = model
+        self.device = device
+
+        # TODO: normalizing_dict is property of the model in crop-mask should this be the same here?
```
Comment (gabrieltseng) [inline crop_classification/inference.py]: yes, this makes sense to me. Where is this loaded though? It seems like currently nothing would set this normalizing_dict to `not None`.

File: crop_classification/tests/test_lstm.py [hunk 3]
```diff
@@ -1,48 +1,70 @@
...
+encoder_state_dict_path = (
+    Path(__file__).parent / "test_data/timl/encoder_state_dict.pth"
+)
 
-def test_jit(tmp_path):
 
-    input_size, num_timesteps, input_task_info_size = 18, 12, 13
-
-    # train a model
+# TODO this might make sense to put into deeplearning.py with more parameters
```
Comment (gabrieltseng) [inline crop_classification/tests/test_lstm.py]: yes. I'd want to get rid of the magic numbers if we did this, maybe by coupling it to the `config` file.

**Review by gabrieltseng — CHANGES_REQUESTED** [review]

File: crop_classification/inference.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    @staticmethod
+    def _tif_to_np(
+        local_path: Path,
+        start_date: datetime,
+        normalizing_dict: Optional[Dict[str, np.ndarray]] = None,
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+
+        # TODO: This is close to Engineer().process_test_file()
+        # but does not rely on the Engineer class being initialized with labels
+        # might make sense to add a majority of this class to cropharvest directly
```
Comment (ivanzvonkov) [inline crop_classification/inference.py]: @gabrieltseng Thoughts about this comment?

File: crop_classification/tests/test_lstm.py (hunk 3 above)
Comment (ivanzvonkov) [reply crop_classification/tests/test_lstm.py]: what do you mean by magic numbers?

File: crop_classification/inference.py (hunk 1 above)
Comment (ivanzvonkov) [reply crop_classification/inference.py]: Good point. I guess batch size during inference should be determined by memory constraints only?
Would using the model training batch size as default make any sense?

File: crop_classification/inference.py (hunk 2 above)
Comment (ivanzvonkov) [reply crop_classification/inference.py]: In `crop-mask` when initializing the `Model` during training I compute (or pull from cache) the normalizing_dict and have an attribute: `self.normalizing_dict_jit: Dict[str, List[float]]`. Since the dictionary only has primitives it can be saved inside the jit model and use when the model is loaded. Would something like that make sense in this repo? How is the normalizing dict computed / used here?

Comment (gabrieltseng) [reply crop_classification/inference.py]: We load the normalizing dict from the `cropharvest` normalizing dict (so its loaded as part of the cropharvest datasets).

I think it would be nice to have a check somehow; by default we expect a normalizing dict, and the model should throw a warning if it doesn't have one

File: crop_classification/tests/test_lstm.py (hunk 3 above)
Comment (gabrieltseng) [reply crop_classification/tests/test_lstm.py]: e.g. `num_classification_layers=2`

File: crop_classification/inference.py (hunk 4 above)
Comment (gabrieltseng) [reply crop_classification/inference.py]: hmm. do you mean coupling the inference class to cropharvest?

I would prefer to keep it here, since its a specific application of the dataset. What might make sense is to move this function into the engineer, as a class method

```python
# (in Engineer)
     @classmethod
     def tif_to_np(cls, local_path, start_date):
         da, slope = cls.load_tif(local_path, start_date)
        ...
```
Then, we can just call that function here, and do the same in `process_test_file`

File: crop_classification/tests/test_lstm.py (hunk 3 above)
Comment (ivanzvonkov) [reply crop_classification/tests/test_lstm.py]: Would `crop_classification/dl/timl.py` make sense for this `load_model` function?

Comment (gabrieltseng) [reply crop_classification/tests/test_lstm.py]: yes, that makes sense to me

**Comment by gabrieltseng** [comment]
Also, just a small note @ivanzvonkov; would you mind adding something small in the crop_classification/README explaining what the inference file is for

File: crop_classification/inference.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+import torch
+
+from typing import Dict, List, Optional, Tuple, Union
+
+
+class Inference:
+    """
+    Class for running inference using either a pytorch checkpoint model or
+    a pytorch jit model on a single tif file.
+    """
```
Comment (ivanzvonkov) [inline crop_classification/inference.py]: @gabrieltseng would adding more detail here (or to the top of the file) be more useful than the readme? It would be more tightly coupled to this file

File: crop_classification/inference.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    ):
+        self.model = model
+        self.device = device
+
+        try:
+            self.normalizing_dict: Dict[str, np.ndarray] = {
+                k: np.array(v) for k, v in self.model.normalizing_dict_jit.items()
+            }
+        except AttributeError:
+            self.normalizing_dict: Dict[str, np.ndarray] = None
```
Comment (ivanzvonkov) [inline crop_classification/inference.py]: Re normalizing dict, let me know if something like this makes sense @gabrieltseng

File: crop_classification/dl/timl.py [hunk 7]
```diff
@@ -717,3 +724,42 @@ def train_timl_model(
...
+        classifier_vector_size=HIDDEN_VECTOR_SIZE,
+    )
+    model.load_state_dict(torch.load(model_state_dict_path))
+
+    if encoder_state_dict_path.exists():
+        encoder = TaskEncoder(
+            input_size=task_info.shape[0],
+            encoder_vector_sizes=ENCODER_VECTOR_SIZES,
+            encoder_dropout=ENCODER_DROPOUT,
+            num_bands=input_size,  # TODO is this right?
```
Comment (ivanzvonkov) [inline crop_classification/dl/timl.py]: This was `num_bands=dataset.num_bands` before but I believe if the encoder is enabled then `input_size` is `dataset.num_bands`. Let me know if ok

File: crop_classification/dl/timl.py [hunk 8]
```diff
@@ -625,13 +632,13 @@ def _make_tasks(
...
-    classifier_base_layers: int,
-    num_classification_layers: int,
     model_name: str,
     encoder_vector_sizes: Union[List[int], int],
-    encoder_dropout: float,
+    classifier_vector_size: int = HIDDEN_VECTOR_SIZE,
+    classifier_dropout: float = CLASSIFIER_DROPOUT,
+    classifier_base_layers: int = CLASSIFIER_BASE_LAYERS,
+    num_classification_layers: int = NUM_CLASSIFICATION_LAYERS,
+    encoder_dropout: float = ENCODER_DROPOUT,
```
Comment (ivanzvonkov) [inline crop_classification/dl/timl.py:641]: Decided to just use the config by default here, any thoughts about this?

File: crop_classification/deep_learning.py [hunk 9]
```diff
@@ -10,28 +10,21 @@
...
     TrainDataLoader,
     concatenate_task_info,
+    load_timl_model,
 )
 
 from typing import Optional
 
 
-def run(data_folder: Path = DATAFOLDER_PATH, zero_shot: bool = False) -> None:
+def evaluate_model(data_folder: Path = DATAFOLDER_PATH, zero_shot: bool = False) -> None:
```
Comment (ivanzvonkov) [inline crop_classification/deep_learning.py]: Thought it might be a little more clear with this name

File: crop_classification/dl/config.py [hunk 10]
```diff
@@ -8,3 +8,12 @@
...
 }
+
+# LSTM model configurations
+HIDDEN_VECTOR_SIZE = 128
+NUM_CLASSIFICATION_LAYERS = 2
+CLASSIFIER_DROPOUT = 0.2
+CLASSIFIER_BASE_LAYERS = 1
+
+ENCODER_VECTOR_SIZES = [32, 64, 128]
+ENCODER_DROPOUT = 0.2
```
Comment (ivanzvonkov) [inline crop_classification/dl/config.py:19]: I moved the config here because it's all relevant to `crop_classification/dl/timl.py` and this makes it easy to import. What do you think about this? Do you think `crop_classification/dl/timl.py` should reach into `crop_classification/config.py` instead? Should it touch the config at all?

File: crop_classification/inference.py (hunk 4 above)
Comment (ivanzvonkov) [reply crop_classification/inference.py]: > do you mean coupling the inference class to cropharvest?

Yes

> I would prefer to keep it here, since its a specific application of the dataset

I think the this `Inference` class actually should work with any application of the dataset. I made an almost identical one for crop-mask. It essentially just simplifies running a cropharvest exported tif through a model trained on cropharvest data.

File: crop_classification/inference.py (hunk 5 above)
Comment (gabrieltseng) [reply crop_classification/inference.py]: yes, that works too. I think a brief thing in the readme would be good as well though (could literally be a single line)

File: crop_classification/dl/timl.py (hunk 8 above)
Comment (gabrieltseng) [reply crop_classification/dl/timl.py:641]: yes this is great

File: crop_classification/deep_learning.py (hunk 9 above)
Comment (gabrieltseng) [reply crop_classification/deep_learning.py]: We should make this change in cropharvest too

File: crop_classification/dl/config.py (hunk 10 above)
Comment (gabrieltseng) [reply crop_classification/dl/config.py:19]: yea this is fine for me. I can see the benefits of both ways, but I like what this allows (e.g. the default args you did)

File: crop_classification/inference.py (hunk 4 above)
Comment (gabrieltseng) [reply crop_classification/inference.py]: hmm, yes actually good point. I think it makes sense for this to be moved into the cropharvest repo then

Comment (ivanzvonkov) [reply crop_classification/inference.py]: Moved in this PR: https://github.com/nasaharvest/cropharvest/pull/75

**Comment by ivanzvonkov** [comment]
@gabrieltseng Updated to use cropharvest's inference class. Let me know if you can take a second look!

**Comment by ivanzvonkov** [comment]
Should pass when cropharvest 0.3.0 is deployed

Note to self: Trigger tests by making a commit to remove #TODO comments

File: crop_classification/dl/timl.py (hunk 7 above)
Comment (gabrieltseng) [reply crop_classification/dl/timl.py]: this works :)

File: crop_classification/tests/test_lstm.py [hunk 11]
```diff
@@ -1,47 +1,34 @@
-import torch
 import numpy as np
+import sys
+import torch
 from pathlib import Path
+from cropharvest.inference import Inference
 
-from dl.lstm import Classifier
-from dl.encoder import TaskEncoder
+# Needed for running pytest without python -m pytest
```
Comment (gabrieltseng) [inline crop_classification/tests/test_lstm.py:7]: Is there a situation where we want to do this?

**Review by gabrieltseng — APPROVED** [review]

File: crop_classification/tests/test_lstm.py (hunk 11 above)
Comment (ivanzvonkov) [reply crop_classification/tests/test_lstm.py:7]: Yes, it allows me to run the tests from inside VSCode. Running them in VSCode makes it possible to add debug points and step through code starting from a test. Haven't figured out how to achieve that any other way

See: https://stackoverflow.com/questions/57273945/imports-break-vscode-testing-with-pytest
